test/functional/retryable_writes.test.js

```diff
@@ -94,6 +93,12 @@ function executeScenarioTest(test, ctx) {
             expect(err.message, 'expected operations to fail, but they succeeded').to.not.match(
               /expected false to be true/
             );
+            const errorLabelsContain =
```

Comment:
I think this is one piece of the puzzle, but the ticket explicitly calls out the bug of us treating the presence of `outcome.error` and `outcome.result` as mutually exclusive. I think that entails just removing the `else` from L103 and any errors that ensue

Reply:
I don't believe it makes sense to remove the check. The promise will still split this logic up. 

I added a check to check if the `result` has the label, and if not, if it's a true result, check `err.result`, this is for bulk writes. 

This change will result in test failures, there is no `err.result` it's a `MongoNetworkError`, i've been looking through the codebase to find where I would be able to change this to a Bulk write error. Any pointers?

Reply:
You're right, we can't just get rid of the else for the reasons you listed above, didn't think about promises here but I think you've captured the gist of my suggestion.

Regarding the bugs you've uncovered, I was able to get around the first set of them with this patch:
```
diff --git a/src/bulk/common.ts b/src/bulk/common.ts
index 99b6fba00..335c426b8 100644
--- a/src/bulk/common.ts
+++ b/src/bulk/common.ts
@@ -550,7 +550,7 @@ function executeCommands(
   function resultHandler(err?: AnyError, result?: Document) {
     // Error is a driver related error not a bulk op error, return early
     if (err && 'message' in err && !(err instanceof MongoWriteConcernError)) {
-      return callback(err);
+      return callback(new BulkWriteError(err, new BulkWriteResult(bulkOperation.s.bulkResult)));
     }
 
     if (err instanceof MongoWriteConcernError) {
diff --git a/test/functional/retryable_writes.test.js b/test/functional/retryable_writes.test.js
index 17c1df860..7daf3870c 100644
--- a/test/functional/retryable_writes.test.js
+++ b/test/functional/retryable_writes.test.js
@@ -98,7 +98,7 @@ function executeScenarioTest(test, ctx) {
               /expected false to be true/
             );
             if (hasResult) {
-              expect(err.result).to.equal(test.outcome.result);
+              expect(err.result).to.matchMongoSpec(test.outcome.result);
             }
             const errorLabelsContain =
               test.outcome && test.outcome.result && test.outcome.result.errorLabelsContain;
```

Two things to note here:
- I'm using the more versatile `matchMongoSpec` `chai` helper, which follows our relatively complicated rules for matching spec results
- The actual issue was that bulk was passing through errors that it probably should be wrapping. This is kind of a tough case that will need more consideration, because theoretically this would return something like a `TypeError` to the user as a `BulkWriteError` - which may be fine, but someone (you!) needs to sit and think about the implications of this change.

When I ran it with that patch everything passed against a single server, and there were three failures against a replicaset which seemed to be mostly in transactions code, and specifically about propagating error labels. That will bear some more investigation as well.